Java: Centralize and model additional path creations.

intrigus-lgtm opened the pull request:
/** Models the [java.nio.Path.resolveSibling](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Path.html#resolveSibling(java.lang.String)) and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Path.html#resolveSibling(java.nio.file.Path) method. */

/** Models the [java.nio.Path.of](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Path.html#of(java.lang.String,java.lang.String...)) and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Path.html#of(java.net.URI) method. */

/** Models the [java.nio.Path.resolve](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Path.html#resolve(java.lang.String)) and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Path.html#resolve(java.nio.file.Path) method. */

/** Models the [new java.io.FileReader(...)](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/FileReader.html) constructor. */

/** Models the [new java.io.FileInputStream(...)](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/FileInputStream.html) constructor. */

/** Models the [new java.io.FileOutputStream(...)](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/FileOutputStream.html) constructor. */

intrigus-lgtm commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+/** Models the creation of a path. */
+abstract class PathCreation extends Expr {
+  /** Gets an input that is used in the creation of this path. */
+  abstract Expr getInput();
```
Should this be changed to `getAnInput`?
Maybe this should also clarify that this will only return inputs that are `String`s, if any?

Marcono1234 replied:
`getAnInput` would probably be better given that for example `Path.of(String, String...)` has multiple inputs.

> Maybe this should also clarify that this will only return inputs that are `String`s, if any?

This is currently not consistently the case, below `Paths.get` currently also matches `Paths.get(URI)`.
And in my opinion at least matching `URI` as well might be useful.

aschackmull replied:
I agree that `getAnInput` is a better name - let's change it.

aschackmull replied:
Regarding the qldoc clarification, perhaps add a line saying something like ``This excludes inputs of type `File` and `Path`.``

intrigus-lgtm commented:
Should methods like [java.nio.file.Files.newBufferedReader](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#newBufferedReader-java.nio.file.Path-) also be added?

Marcono1234 commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+/**
+ * Models the different ways to create paths. Either by using `java.io.File`-related APIs or `java.nio.Path`-related APIs.
```
```suggestion
 * Models the different ways to create paths. Either by using `java.io.File`-related APIs or `java.nio.file.Path`-related APIs.
```

Marcono1234 commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+  }
+}
+
+/** Models the `java.nio.Path.resolve` method. */
```
```suggestion
/** Models the `java.nio.file.Path.resolve` method. */
```

Marcono1234 commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+  }
+}
+
+/** Models the `java.nio.Path.resolveSibling` method. */
```
```suggestion
/** Models the `java.nio.file.Path.resolveSibling` method. */
```

Marcono1234 commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+    def.getName() = "startsWith" or
+    def.getName() = "endsWith" or
+    def.getName() = "isEmpty" or
+    def.getName() = "equals"
```
Isn't an `equals` check strong? E.g.:
```java
if (s.equals("C:/test")) {
    Path.of(s);
}
```

intrigus-lgtm replied:
I didn't create this, so I don't know.

aschackmull replied:
This isn't a particularly precise set of predicates - the filtering here is fairly ad-hoc, but it does indeed also match negated checks.

Marcono1234 commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+  }
+}
+
+/** Models the `java.nio.Path.of` method. */
```
```suggestion
/** Models the `java.nio.file.Path.of` method. */
```

intrigus-lgtm commented:
Any way to retrigger a build?
The check failed for no good reason it seems...

intrigus-lgtm commented:
Rebased this commit onto latest master and now the code scanning stuff seems to work, so this should be ready for review now :)

aschackmull commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+}
+
+/** Models the `java.nio.file.Paths.get` method. */
+class PathsGet extends PathCreation, MethodAccess {
```
I think perhaps all of the subclasses of `PathCreation` ought to be `private`.

aschackmull commented on java/ql/test/library-tests/pathcreation/PathCreation.ql:
```diff
@@ -0,0 +1,5 @@
+import semmle.code.java.security.PathCreation
+
+from PathCreation path
+select path
```
```suggestion
select path, path.getAnInput()
```
This makes the test a bit nicer and more complete.

aschackmull commented on java/ql/src/semmle/code/java/security/PathCreation.qll:
```diff
@@ -0,0 +1,165 @@
+
+// Ignore cases where the variable has been checked somehow,
+// but allow some particularly obviously bad cases.
+predicate guarded(VarAccess e) {
```
I think the two predicates `inWeakCheck` and `guarded` are sufficiently ad-hoc that we probably want to keep them next to the queries where they aren't as exposed - they are not worth exporting here for more general import into other queries.

intrigus-lgtm replied:
I wanted to use them in https://github.com/github/codeql/pull/3794 but the predicates indeed seem to not be too general purpose.
And the linked query will go to experimental first anyway so I will move the two predicates back.

intrigus-lgtm commented:
I hope I've addressed all your review comments.


I only have this question left for you:
> Should methods like [java.nio.file.Files.newBufferedReader](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#newBufferedReader-java.nio.file.Path-) also be added?

Previously a similar instance was already modeled.
https://github.com/github/codeql/blob/e167b871502966059cd45b6e3258bfa70062d2c9/java/ql/src/Security/CWE/CWE-022/PathsCommon.qll#L40-L48

aschackmull commented:
> > Should methods like [java.nio.file.Files.newBufferedReader](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#newBufferedReader-java.nio.file.Path-) also be added?
> 
> Previously a similar instance was already modeled.

No, `newBufferedReader` only takes `Path` input, whereas `FileWriter` accepts a plain string and treats it as a filename, and the library aims to capture the points where strings are interpreted as names of paths/files.

I believe the unit test output needs an update, since an additional column was added to the test, but isn't yet reflected in the `.expected` file.  Otherwise this LGTM.

intrigus-lgtm commented:
> I believe the unit test output needs an update, since an additional column was added to the test, but isn't yet reflected in the `.expected` file. Otherwise this LGTM.

Fixed
